src/Cli/dotnet/Commands/Run/Api/RunApiCommand.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.CommandLine;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.DotNet.Cli.Commands.Run.Api;
+
+/// <summary>
+/// Takes JSON from stdin lines, produces JSON on stdout lines, doesn't perform any changes.
+/// Can be used by IDEs to see the project file behind a file-based program.
+/// </summary>
+internal sealed class RunApiCommand(ParseResult parseResult) : CommandBase(parseResult)
+{
+    public override int Execute()
+    {
+        for (string? line; (line = Console.ReadLine()) != null;)
+        {
+            if (string.IsNullOrWhiteSpace(line))
+            {
+                continue;
+            }
+
+            try
+            {
+                RunApiInput input = JsonSerializer.Deserialize(line, RunFileApiJsonSerializerContext.Default.RunApiInput)!;
+                RunApiOutput output = input.Execute();
+                Respond(output);
+            }
+            catch (Exception ex)
+            {
+                Respond(new RunApiOutput.Error { Message = ex.Message, Details = ex.ToString() });
+            }
+        }
+
+        return 0;
+
+        static void Respond(RunApiOutput message)
+        {
+            string json = JsonSerializer.Serialize(message, RunFileApiJsonSerializerContext.Default.RunApiOutput);
+            Console.WriteLine(json);
+        }
+    }
+}
+
+[JsonDerivedType(typeof(GetProject), nameof(GetProject))]
+internal abstract class RunApiInput
+{
+    private RunApiInput() { }
+
+    public abstract RunApiOutput Execute();
+
+    public sealed class GetProject : RunApiInput
+    {
+        public string? ArtifactsPath { get; init; }
```

Comment:
What it does mean for us to pass this in? I would have imagined you were returning it?

Reply:
Yes, this is optional, the SDK normally computes it, but you can overwrite it which is useful for tests. I will add a clarifying comment.